google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java

```diff
@@ -0,0 +1,341 @@
+
+              @Override
+              public void onResponse(Row row) {
+                try {
+                  counter.incrementAndGet();
+                  Thread.sleep(applicationLatency);
+                } catch (InterruptedException e) {
+                }
+              }
+
+              @Override
+              public void onError(Throwable throwable) {
+                future.setException(throwable);
+              }
+
+              @Override
+              public void onComplete() {
+                future.set(null);
+              }
+            });
+    future.get();
+
+    verify(statsRecorderWrapper).putApplicationLatencies(longValue.capture());
+
+    assertThat(counter.get()).isGreaterThan(0);
+    assertThat(longValue.getValue()).isAtLeast(applicationLatency * counter.get());
+    assertThat(longValue.getValue()).isLessThan(applicationLatency * (counter.get() + 1));
+  }
+
+  @Test
+  public void testRetryCount() {
+    stub.mutateRowCallable()
+        .call(RowMutation.create(TABLE_ID, "random-row").setCell("cf", "q", "value"));
+
+    verify(statsRecorderWrapper).putRetryCount(intValue.capture());
+
+    assertThat(intValue.getValue()).isEqualTo(3);
+  }
+
+  @Test
+  public void testMutateRowAttempts() {
+    stub.mutateRowCallable()
+        .call(RowMutation.create(TABLE_ID, "random-row").setCell("cf", "q", "value"));
+
+    verify(statsRecorderWrapper, times(4))
+        .record(status.capture(), tableId.capture(), zone.capture(), cluster.capture());
+    assertThat(zone.getAllValues().get(0)).isEqualTo(UNDEFINED);
+    assertThat(zone.getAllValues().get(1)).isEqualTo(UNDEFINED);
+    assertThat(zone.getAllValues().get(2)).isEqualTo(UNDEFINED);
+    assertThat(cluster.getAllValues().get(0)).isEqualTo(UNDEFINED);
+    assertThat(cluster.getAllValues().get(1)).isEqualTo(UNDEFINED);
+    assertThat(cluster.getAllValues().get(2)).isEqualTo(UNDEFINED);
+    assertThat(status.getAllValues().get(0)).isEqualTo("UNAVAILABLE");
+    assertThat(status.getAllValues().get(1)).isEqualTo("UNAVAILABLE");
+    assertThat(status.getAllValues().get(2)).isEqualTo("OK");
+  }
+
+  private class FakeService extends BigtableGrpc.BigtableImplBase {
+
+    @Override
+    public void readRows(
+        ReadRowsRequest request, StreamObserver<ReadRowsResponse> responseObserver) {
+      try {
+        Thread.sleep(SERVER_LATENCY);
+      } catch (InterruptedException e) {
+      }
+      responseObserver.onNext(READ_ROWS_RESPONSE_1);
+      responseObserver.onNext(READ_ROWS_RESPONSE_2);
+      responseObserver.onNext(READ_ROWS_RESPONSE_3);
+      responseObserver.onCompleted();
+    }
+
+    @Override
+    public void mutateRow(
+        MutateRowRequest request, StreamObserver<MutateRowResponse> responseObserver) {
+      if (serverRetryDelayStopwatch.isRunning()) {
+        serverTotalRetryDelay.addAndGet(serverRetryDelayStopwatch.elapsed(TimeUnit.MILLISECONDS));
+        serverRetryDelayStopwatch.reset();
+      }
```

Comment:
I think you might cache coherences issues here: the thread that create serverRetryDelayStopwatch can be different from the current one and the values might not have been flushed to memory